tools/SampleAnalyzer/Process/Counter/Counter.h

```diff
@@ -1,114 +1,128 @@
-  Counter(const std::string& name = "unkwown")
-  { 
-    name_       = name;
-    nentries_   = std::make_pair(0,0); 
-    sumweight_  = std::make_pair(0.,0.);
-    sumweight2_ = std::make_pair(0.,0.);
-  }
-
-  /// Destructor
-  ~Counter()
-  { }
-
-  /// Reset
-  void Reset()
-  {
-    nentries_   = std::make_pair(0,0); 
-    sumweight_  = std::make_pair(0.,0.);
-    sumweight2_ = std::make_pair(0.,0.);
-  }
-
-  /// Increment the counter
-  void Increment(const MAfloat32& weight=1.)
-  {
-    if (weight>0)
-    {
-      nentries_.first++;
-      sumweight_.first+=weight;
-      sumweight2_.first+=weight*weight;
-    }
-    else if (weight<0)
+    class Counter
     {
-      nentries_.second++;
-      sumweight_.second+=weight;
-      sumweight2_.second+=weight*weight;
-    }
-  }
-
-};
+        friend class CounterManager;
+
+        // -------------------------------------------------------------
+        //                        data members
+        // -------------------------------------------------------------
+    public:
+        /// name of the analysis
+        std::string name_;
+
+        /// number of times the function Increment is called
+        /// first = positive weight ; second = negative weight
+        std::map<MAint32, ENTRIES> nentries_;
+
+        /// sum of weights
+        /// first = positive weight ; second = negative weight
+        std::map<MAint32, WEIGHTS> sumweights_;
+
+        /// sum of squared weights
+        /// first = positive weight ; second = negative weight
+        std::map<MAint32, WEIGHTS> sumweights2_;
+
+        // -------------------------------------------------------------
+        //                       method members
+        // -------------------------------------------------------------
+    public:
+        /// Constructor without argument
+        Counter(const std::string &name = "unkwown")
+        {
+            name_ = name;
+            Reset();
+        }
+
+        /// Destructor
+        ~Counter() {}
+
+        /// Reset
+        void Reset()
+        {
+            nentries_.clear();
+            sumweights_.clear();
+            sumweights2_.clear();
+        }
+
+        MAint32 size() { return nentries_.size(); }
+
+        void Initialise(const WeightCollection &multiweight)
+        {
+            Reset();
+            for (auto &weight : multiweight.GetWeights())
+            {
+
+                nentries_[weight.first] = ENTRIES();
+                sumweights_[weight.first] = WEIGHTS();
+                sumweights2_[weight.first] = WEIGHTS();
+            }
+        }
+
+        std::map<MAint32, ENTRIES> nentries() { return nentries_; }
+        std::map<MAint32, WEIGHTS> sumW() { return sumweights_; }
+        std::map<MAint32, WEIGHTS> sumW2() { return sumweights2_; }
+
+        /// Increment the counter
+        void Increment(const WeightCollection &multiweight)
+        {
+            for (auto &weight : multiweight.GetWeights())
```

Comment:
The filling of the `nentries_` element has to be done outside the loop.

Reply:
Actually this is tricky. Weights can be positive and negative and there is nothing stopping them to be positive for this weight and negative for the other one so I think having entries separate for every weight makes sense.

Reply:
ooh I didn't think about this... Then forget everything I wrote about entries.